File: lib/data-structures/SingleRouteCandidatePriorityQueue.ts
export type Node = {
  x: number
  y: number
  z: number

  g: number
  h: number
  f: number

  parent: Node | null
}

export class SingleRouteCandidatePriorityQueue<T extends Node = Node> {
  private heap: T[] = []
  private priorities: Float64Array | null

  constructor(nodes: T[], options?: { cachePriorities?: boolean }) {
    this.heap = []
    // Routing candidates have fixed costs after enqueueing. Keep those costs
    // contiguous so heap comparisons do not chase references to search nodes.
    this.priorities = options?.cachePriorities ? new Float64Array(64) : null

    for (const node of nodes) {
      this.enqueue(node)
    }
  }

  // Removing an element will remove the
  // top element with highest priority then
  // heapifyDown will be called
  dequeue(): T | null {
    if (this.heap.length === 0) {
      return null
    }
    const item = this.heap[0]
    this.heap[0] = this.heap[this.heap.length - 1]
    if (this.priorities) {
      this.priorities[0] = this.priorities[this.heap.length - 1]
    }
    this.heap.pop()
    this.heapifyDown()
    return item
  }

  peek(): T | null {
    if (this.heap.length === 0) {
      return null
    }
    return this.heap[0]
  }

  enqueue(item: T) {
    if (this.priorities) {
      if (this.heap.length === this.priorities.length) {
        const priorities = new Float64Array(this.priorities.length * 2)
        priorities.set(this.priorities)
        this.priorities = priorities
      }
      this.priorities[this.heap.length] = item.f
    }
    this.heap.push(item)
    this.heapifyUp()
  }

  heapifyUp() {
    let index = this.heap.length - 1
    const item = this.heap[index]
    const priorities = this.priorities
    const itemPriority = priorities ? priorities[index] : item.f
    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2)
      const parent = this.heap[parentIndex]
      const parentPriority = priorities ? priorities[parentIndex] : parent.f
      if (parentPriority <= itemPriority) break
      this.heap[index] = parent
      if (priorities) priorities[index] = parentPriority
      index = parentIndex
    }
    this.heap[index] = item
    if (priorities) priorities[index] = itemPriority
  }

  heapifyDown() {
    let index = 0
    const heapLength = this.heap.length
    const item = this.heap[index]
    if (!item) return
    const priorities = this.priorities
    const itemPriority = priorities ? priorities[index] : item.f
    while (true) {
      const leftChildIndex = 2 * index + 1
      if (leftChildIndex >= heapLength) break
      const rightChildIndex = leftChildIndex + 1
      let smallerChildIndex = leftChildIndex
      const leftPriority = priorities
        ? priorities[leftChildIndex]
        : this.heap[leftChildIndex].f
      if (
        rightChildIndex < heapLength &&
        (priorities ? priorities[rightChildIndex] : this.heap[rightChildIndex].f) <
          leftPriority
      ) {
        smallerChildIndex = rightChildIndex
      }
      const childPriority = priorities
        ? priorities[smallerChildIndex]
        : this.heap[smallerChildIndex].f
      if (itemPriority < childPriority) {
        break
      }
      this.heap[index] = this.heap[smallerChildIndex]
      if (priorities) priorities[index] = childPriority
      index = smallerChildIndex
    }
    this.heap[index] = item
    if (priorities) priorities[index] = itemPriority
  }

  /**
   * Returns the top N candidates sorted by f value (lowest first)
   */
  getTopN(n: number): T[] {
    return [...this.heap].sort((a, b) => a.f - b.f).slice(0, n)
  }
}

File: rust/general-router/src/single_route_candidate_priority_queue.rs
use std::rc::Rc;

#[derive(Clone, Debug)]
pub struct Node {
    pub x: f64,
    pub y: f64,
    pub z: f64,
    pub g: f64,
    pub h: f64,
    pub f: f64,
    pub parent: Option<Rc<Node>>,
}

#[derive(Clone, Debug, Default)]
pub struct SingleRouteCandidatePriorityQueue {
    heap: Vec<Rc<Node>>,
}

impl SingleRouteCandidatePriorityQueue {
    pub fn new(nodes: Vec<Rc<Node>>) -> Self {
        let mut queue = Self { heap: Vec::new() };
        for node in nodes {
            queue.enqueue(node);
        }
        queue
    }

    pub fn dequeue(&mut self) -> Option<Rc<Node>> {
        if self.heap.is_empty() {
            return None;
        }
        let item = self.heap[0].clone();
        self.heap[0] = self.heap[self.heap.len() - 1].clone();
        self.heap.pop();
        self.heapify_down();
        Some(item)
    }

    pub fn peek(&self) -> Option<Rc<Node>> {
        if self.heap.is_empty() {
            return None;
        }
        Some(self.heap[0].clone())
    }

    pub fn enqueue(&mut self, item: Rc<Node>) {
        self.heap.push(item);
        self.heapify_up();
    }

    pub fn heapify_up(&mut self) {
        let mut index = self.heap.len() - 1;
        let item = self.heap[index].clone();
        while index > 0 {
            let parent_index = (index - 1) / 2;
            let parent = self.heap[parent_index].clone();
            if parent.f <= item.f {
                break;
            }
            self.heap[index] = parent;
            index = parent_index;
        }
        self.heap[index] = item;
    }

    pub fn heapify_down(&mut self) {
        let mut index = 0;
        let heap_length = self.heap.len();
        let Some(item) = self.heap.first().cloned() else { return; };
        loop {
            let left_child_index = 2 * index + 1;
            if left_child_index >= heap_length {
                break;
            }
            let right_child_index = left_child_index + 1;
            let mut smaller_child_index = left_child_index;
            if right_child_index < heap_length
                && self.heap[right_child_index].f < self.heap[left_child_index].f
            {
                smaller_child_index = right_child_index;
            }
            if item.f < self.heap[smaller_child_index].f {
                break;
            }
            self.heap[index] = self.heap[smaller_child_index].clone();
            index = smaller_child_index;
        }
        self.heap[index] = item;
    }

    pub fn get_top_n(&self, n: usize) -> Vec<Rc<Node>> {
        let mut candidates = self.heap.clone();
        candidates.sort_by(|a, b| (a.f - b.f).partial_cmp(&0.0).unwrap_or(std::cmp::Ordering::Equal));
        candidates.truncate(n);
        candidates
    }
}
